File: python-gdb-activities-20/activity19-solution/gdb/command/withdraw_command.py
# gdb/command/withdraw_command.py
from typing import Optional
from gdb.domain.iaccount import IAccount
from gdb.domain.transaction import Transaction
from gdb.command.transaction_command import TransactionCommand

class WithdrawCommand(TransactionCommand):
    """Command executing a withdrawal operation."""

    def __init__(self, account: IAccount, amount: float, pin: int) -> None:
        self._account = account
        self._amount = float(amount)
        self._pin = int(pin)
        self._transaction: Optional[Transaction] = None

    def execute(self) -> None:
        if hasattr(self._account, "withdraw_with_transaction"):
            self._transaction = self._account.withdraw_with_transaction(self._amount, self._pin)
        else:
            self._account.withdraw(self._amount, self._pin)

    def get_transaction(self) -> Optional[Transaction]:
        return self._transaction

    def getTransaction(self) -> Optional[Transaction]:
        return self.get_transaction()

File: java-gdb-activities-20/activity18/src/com/gdb/command/WithdrawCommand.java
package com.gdb.command;

import com.gdb.domain.Account;
import com.gdb.domain.IAccount;
import com.gdb.domain.Transaction;

public class WithdrawCommand implements TransactionCommand {
    private static final long serialVersionUID = 1L;

    private IAccount account;
    private double amount;
    private int pin;
    private Transaction transaction;

    public WithdrawCommand(IAccount account, double amount, int pin) {
        this.account = account;
        this.amount = amount;
        this.pin = pin;
    }

    @Override
    public void execute() throws Exception {
        if (account instanceof Account) {
            this.transaction = ((Account) account).withdrawWithTransaction(amount, pin);
        } else {
            account.withdraw(amount, pin);
        }
    }

    @Override
    public Transaction getTransaction() {
        return this.transaction;
    }
}
